rfcs/unboxed-types.md

```diff
@@ -0,0 +1,458 @@
+    type #int32 : int32
+
+All of these can be passed and returned from functions, but only the
+`value` types (including `immediate` ones) can be stored in, say,
+lists. Note that the existing `+.`, etc. primitives won't work on
+`#float`, because they have the wrong type. New ones for `#float` will
+be made available. (Same for `#int32`, etc.)
+
+Unboxed types will be allowed in record fields:
+
+    type t = {
+      foo : int;
+      bar : #float;
+      ...
+    }
+
+Because of how the OCaml runtime works, there are some restrictions on
+how boxed and unboxed types can be mixed in a record. If there are any
+fields which might contain pointers (that is, are of layout `value`
+but are not `immediate`), then all fields must be of layout
+`value`. So, you can mix `int` and `#float` in a single record, but
+not `string` and `#float`.
+
+## Unboxed tuples and records
+
+The second part of this proposal is to allow user-defined unboxed
+types (by contrast to what's above: user-defined boxed types
+containing unboxed fields). The simplest of these is unboxed tuples:
+
+    #( int * float * string )
+
+This is an unboxed tuple, and is passed around as three separate
+values.  Passing it to a function requires three registers.
+The layout of an unboxed tuple is the product of the layout of its
+components. In particular, the layout of `#( string * string )` is
+`value * value`, not `value`, and so it cannot be stored in, say, a
+list.
+
+Unboxed records are in theory just as straightforward, but have some
+extra complexity just to handle the various ways that types are
+defined and used. Suppose you write:
+
+    type t = { x : int32; y : int32 }
+
+Sometimes we want to refer to `t` as a boxed type, to store it in
+lists and other standard data structures. Sometimes we want to refer
+to `t` as an unboxed type, to avoid extra allocations (in particular,
+to inline it into other records). The language never infers which is
+which (boxing and allocation is always explicit in this design), so we
+need different syntax for both.  So, `type t = { ... }` defines *two*
+types, the boxed record `t : value` and the unboxed record `#t :
+int32 * int32`. They have different syntax for introduction and
+elimination:
+
+    let p : t = { x = a; y = b }
+    let px = p.x
+    let q : #t = #{ x = a; y = b }
+    let qx = q.#x
+
+An unboxed record like `q` cannot itself contain mutable
+fields. However, a boxed record can contain a mutable field of type
+`q`. This can be used to inline one record inside another, e.g.:
+
+    type s = { foo : int; mutable bar : #t }
+
+This defines a boxed record, layed out as three fields. We will
+support efficient updates of sub-parts of `bar`, at least with the
+following syntax:
+
+    x.bar <- { x.bar with y = y' }
+
+It would be nice to also allow the syntax `x.bar.#y <- y'`, although
+this is a little trickier to understand and implement, since it is
+`x.bar` rather than `x.bar.y` which is the mutable field.
```

Comment:
@gasche says

> I don't think this is a good idea, due to the awkard confusion with the meaning of x.bar.y <- y' in the non-boxed case.

@stedolan says

> I agree, I'm not terribly happy with this either. It would be nice to have a more compact syntax than x.bar <- { x.bar with foo = 1 } for updating a single field of an inlined record, though.